### PR TITLE
Document and test packed token inputs for MoE and QMoE

### PR DESCRIPTION
**Pull request by apsonawane**
## Description

Clarify and test packed token-major input support for the `MoE` and `QMoE`
operators on the CPU and CUDA execution providers.

Packed inputs use:

- Input: `[total_tokens, hidden_size]`
- Router probabilities: `[total_tokens, num_experts]`
- Output: `[total_tokens, hidden_size]`

Tokens from variable-length sequences can be concatenated without padding
because MoE routing is token-local and does not require sequence boundaries.

## Changes

- Update the `MoE` and `QMoE` schema documentation.
- Document packed input behavior for CPU and CUDA.
- Add CPU parity tests for packed `MoE` and INT4 `QMoE`.
- Add CUDA parity tests for packed `MoE` and INT4 `QMoE`.
- Fix the CPU test harness to pass processed routing probabilities to `MoE`
  while preserving raw router logits for `QMoE`.

No runtime kernel changes are required because the existing CPU and CUDA
implementations already process rank-2 token-major inputs.

### CODE REVIEW

File: onnxruntime/test/python/transformers/test_qmoe_cuda.py [hunk 1]
```diff
@@ -1575,6 +1549,23 @@ def _run_qmoe_cutlass_gemm_second_scale_row_regression(test_case, quant_bits, us
...
+        packed_moe = PhiMoESparseMoeBlock(
+            config,
+            batch_size=1,
+            sequence_length=7,
+            quant_bits=quant_bits,
+            onnx_dtype=TensorProto.FLOAT16,
+            use_asymmetric_quant=False,
+        )
+
+        packed_moe.parity_check()
```
Comment (kunal-vaishnavi) [inline onnxruntime/test/python/transformers/test_qmoe_cuda.py:1567]: From Copilot:

The `quant_bits=0` case does not reliably test the `MoE` graph created by the constructor. `parity_check()` immediately calls `recreate_onnx_model()`, which currently passes `use_quant=True` unconditionally and attempts to emit a `QMoE` model with `expert_weight_bits=0`.

Could we preserve the constructor-created session for this test, as the CPU test does—assert that `packed_moe.ort_sess` is not `None` and call `parity_check(recreate_model=False)`? Alternatively, `recreate_onnx_model()` should branch on `self.quant_bits > 0`. Otherwise, the new CUDA case may fail during model recreation or test a different operator than the intended packed `MoE`.

File: onnxruntime/contrib_ops/cpu/moe/moe_cpu.cc [hunk 2]
```diff
@@ -322,17 +317,31 @@ Status MoE<T>::ComputeMoE(const OpKernelContext* context,
...
+    const size_t token_count = expert_token_map[static_cast<size_t>(expert_idx)].size();
+    if (token_count > 0) {
+      expert_workload.emplace_back(expert_idx, token_count);
+    }
+  }
+  std::sort(expert_workload.begin(), expert_workload.end(),
+            [](const auto& lhs, const auto& rhs) { return lhs.second > rhs.second; });
+
+  std::vector<std::vector<int64_t>> expert_batches(static_cast<size_t>(num_expert_threads));
+  for (size_t i = 0; i < expert_workload.size(); ++i) {
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/moe/moe_cpu.cc:332]: From Copilot:

Sorting experts by token count is helpful, but assigning them round-robin does not necessarily balance the total work per thread. For example, workloads `[8, 7, 6, 5, 4]` produce an `18/12` split across two threads, whereas assigning each expert to the currently least-loaded thread gives `15/15`.

Could we use a least-loaded-thread assignment and add a skewed-routing benchmark to demonstrate the benefit? Otherwise, this unrelated scheduling optimization would be better handled in a dedicated performance PR.

File: onnxruntime/contrib_ops/cuda/llm/moe_gemm/moe_kernels.cu [hunk 3]
```diff
@@ -561,6 +561,12 @@ bool fusedBuildExpertMapsSortFirstToken(const int* token_selected_experts, int*
   // We need enough bits to represent [0, num_experts_per_node+1] (inclusive) i.e. num_experts_per_node + 2 values
   // This is floor(log2(num_experts_per_node+1)) + 1
   int expert_log = static_cast<int>(log2(num_experts_per_node + 1)) + 1;
+  if (expert_log == 10 && experts_per_token == 10) {
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/llm/moe_gemm/moe_kernels.cu:564]: From Copilot:

This enables a new specialized fused path for `expert_log == 10 && experts_per_token == 10`, but the PR does not add a corresponding top-10/high-expert-count parity test.

Could we either remove this unrelated dispatch change or add targeted coverage comparing it with the fallback path, including the 32/64/128/256-thread token-count boundaries and the resource-check fallback? Without that coverage, this PR changes runtime behavior for a previously unsupported fused configuration without validating its permutation maps or output parity.

**Review by titaiwangms — APPROVED** [review]
The documented packed-token contract is consistent with the existing CPU/CUDA input validation, so I am approving.

Non-blocking test suggestion: the new `test_packed_token_input_*` cases still use the same flattened 2-D input path as the existing parity tests. A stronger regression test would compare padded 3-D input with concatenated packed input, ideally using sequences with different lengths, and assert token-for-token output parity. It would also be useful to retain coverage of `normalize_routing_weights=0` in at least one case.
